python_modules/dagster/dagster/_core/definitions/asset_graph_differ.py

```diff
@@ -28,6 +27,79 @@ class AssetDefinitionChangeType(Enum):
     REMOVED = "REMOVED"
 
 
+@whitelist_for_serdes
+@record
+class AssetDefinitionChangeDetailNew:
+    change_type: AssetDefinitionChangeType = AssetDefinitionChangeType.NEW
+
+
+@whitelist_for_serdes
+@record
+class AssetDefinitionChangeDetailCodeVersion:
+    change_type: AssetDefinitionChangeType = AssetDefinitionChangeType.CODE_VERSION
+    base_value: Optional[str] = None
+    branch_value: Optional[str] = None
+
+
+@whitelist_for_serdes
+@record
+class AssetDefinitionChangeDetailDependencies:
+    change_type: AssetDefinitionChangeType = AssetDefinitionChangeType.DEPENDENCIES
+    added: Sequence[AssetKey] = []
+    changed: Sequence[AssetKey] = []
+    removed: Sequence[AssetKey] = []
+
+
+@whitelist_for_serdes
+@record
+class AssetDefinitionChangeDetailPartitionsDefinition:
+    change_type: AssetDefinitionChangeType = AssetDefinitionChangeType.PARTITIONS_DEFINITION
+    base_value: Optional[str] = None
+    branch_value: Optional[str] = None
+
+
+@whitelist_for_serdes
+@record
+class AssetDefinitionChangeDetailTags:
+    change_type: AssetDefinitionChangeType = AssetDefinitionChangeType.TAGS
+    added: Sequence[str] = []
```

Comment:
We generally avoid `[]` as a default value, because the references for all instances of the class point to the same object, so mutating it could lead to very confusing errors. Less of an issue with the `Sequence` type, because it should be immutable, but still a convention we follow.

Reply:
Was curious if dataclasses or records run into this issue - dataclasses explicitly forbid default `[]` values (and force you to use a factory), but record somehow does not display this behavior - not sure why:

```python
from typing import List

from dagster._record import record


class MyNonRecord:
    my_list: List[str] = []


@record
class MyRecord:
    my_list: List[str] = []


a = MyNonRecord()
b = MyNonRecord()

a.my_list.append("hello")
print(a.my_list)
b.my_list.append("world")
print(b.my_list)


c = MyRecord()
d = MyRecord()
c.my_list.append("hello")
print(c.my_list)
d.my_list.append("world")
print(d.my_list)

```


```
['hello']
['hello', 'world']
['hello']
['world']
```

Reply:
(not to say we should get in the habit of using `[]` as a default value 😄 )